domains/pallets/auto-id/src/lib.rs

```diff
@@ -0,0 +1,519 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Pallet AutoID
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+mod tests;
+
+extern crate alloc;
+
+#[cfg(not(feature = "std"))]
+use alloc::collections::BTreeSet;
+use codec::{Decode, Encode};
+use frame_support::dispatch::DispatchResult;
+use frame_support::traits::Time;
+use frame_support::{ensure, PalletError};
+pub use pallet::*;
+use scale_info::TypeInfo;
+use sp_certificate_registry::signature_verification_runtime_interface::verify_signature;
+use sp_certificate_registry::{DerVec, SignatureVerificationRequest};
+use sp_core::U256;
+#[cfg(feature = "std")]
+use std::collections::BTreeSet;
+use subspace_runtime_primitives::Moment;
+use x509_parser::certificate::TbsCertificate;
+use x509_parser::prelude::FromDer;
+
+/// Unique AutoId identifier.
+pub type Identifier = U256;
+
+/// Validity of a given certificate.
+#[derive(Debug, Decode, Encode, TypeInfo, PartialEq, Eq, Clone)]
+pub struct Validity {
+    /// Not valid before the time since UNIX_EPOCH
+    pub not_before: Moment,
+    /// Not valid after the time since UNIX_EPOCH
+    pub not_after: Moment,
+}
+
+impl Validity {
+    /// Checks if the certificate is valid at this time.
+    pub fn is_valid_at(&self, time: Moment) -> bool {
+        time >= self.not_before && time <= self.not_after
+    }
+}
+
+/// Validity conversion error.
+#[derive(TypeInfo, Encode, Decode, PalletError, Debug, PartialEq)]
+pub enum ValidityError {
+    Overflow,
+    Expired,
+}
+
+impl<T: Config> From<ValidityError> for Error<T> {
+    fn from(value: ValidityError) -> Self {
+        Error::<T>::InvalidValidity(value)
+    }
+}
+
+impl TryFrom<x509_parser::prelude::Validity> for Validity {
+    type Error = ValidityError;
+
+    fn try_from(value: x509_parser::certificate::Validity) -> Result<Self, Self::Error> {
+        Ok(Validity {
+            not_before: (value.not_before.timestamp() as u64)
+                .checked_mul(1000)
+                .and_then(|secs| {
+                    secs.checked_add(value.not_before.to_datetime().millisecond() as u64)
+                })
+                .ok_or(Self::Error::Overflow)?,
+            not_after: (value.not_after.timestamp() as u64)
+                .checked_mul(1000)
+                .and_then(|secs| {
+                    secs.checked_add(value.not_after.to_datetime().millisecond() as u64)
+                })
+                .ok_or(Self::Error::Overflow)?,
+        })
+    }
+}
+
+/// Root X509 Certificate.
+#[derive(Debug, Decode, Encode, TypeInfo, PartialEq, Eq, Clone)]
+pub struct X509CertificateRoot {
+    /// Serial number for this certificate
+    pub serial: U256,
+    /// Der encoded certificate's subject.
+    pub subject: DerVec,
+    /// Der encoded certificate's subject's public key info
+    pub subject_pki: DerVec,
```

Comment:
hmmm interesting. Within the context of X509 cert from RFC, PKI always represents public key info. So I did not feel necessary to expand since the field is within the context of X509 struct.
Can change this if you have a strong opinion

Reply:
I think it will be clearer if we expand to use `public_key_info` rather than `pki`.